content/blog/2022/2022-02-14-spring-shell.md

```diff
@@ -0,0 +1,295 @@
+```java
+@ShellMethod(value = "sign out")  
+@ShellMethodAvailability("signOutCheck")  
+public void signOut()  
+{  
+    this.signedIn = false;  
+    log.info("Signed out!");  
+}
+ 
+public Availability signOutCheck()  
+{  
+    return signedIn ?  
+            Availability.available() : Availability.unavailable("Must be signed in first");  
+}
+``` 
+The **third** and final way is to have several methods attached to a single availability method.
+We are going to use the annotation `ShellMethodAvailability` with an array of the **commands names (not method names)**:
+
+```java
+@ShellMethod(value = "sign out")  
+public void signOut()  
+{  
+    this.signedIn = false;  
+    log.info("Signed out!");  
+}  
+  
+@ShellMethod(value = "Change password")  
+public void changePass(@ShellOption String newPass)  
+{  
+    log.info(format("Changed password to '%s'", newPass));  
+}  
+  
+@ShellMethodAvailability({"sign-out", "change-pass"})  
+public Availability signOutCheck()  
+{  
+    return signedIn ?  
+            Availability.available() : Availability.unavailable("Must be signed in first");  
+}
+```
+
+## Other Cool Features in Spring Shell
+Since Spring Shell builds on top of [JLine](https://github.com/jline/jline3) it inherits a lot of its features. Let's look at some of them:
+
+### Tab Completion
+
+### Built in Commands
+
+- help:
+  - lists all the commands known to the shell, including the built-in commands and commands we wrote.
+
+- script:
+  - accepts a local file as an argument and will replay commands found there, one at a time.
+
+
+## Styling the Shell
+We can do so by registering a bean of type `PromptProvider` which includes information on how to render the Shell prompt.
+For example, let's change the prompt text to `Cool Machine==> ` with a green color for the text:
+```java
+@Component  
+public class CustomPromptProvider implements PromptProvider {  
+    @Override  
+  public AttributedString getPrompt() {  
+            return new AttributedString(  
+                    "Cool Machine" + "==> ",  
+                    AttributedStyle.DEFAULT.background(AttributedStyle.GREEN));  
+    }  
+}
+```
+
+## Running Spring Shell Jar File
+To build a JAR file from our Spring Shell project we have two paths depending on the building tool we are using, let's explore them both.
+Please note that we are doing this using command line instructions. If you are using an IDE it will take care of this for you when running the project.
+
+### Maven
+Using Maven we have to run the command `mvn clean package` in our root directory (the directory with the `pom.xml` file). This will produce a JAR file in the `build` directory.
+
+
+### Gradle
+As for Gradle we have to invoke the build task. Using Windows the command will be `./gradlew build`. This will also build the JAR file in the `build` directory.
```

Comment:
I think we can remove the Maven and Gradle sections. This article is not about how to create a JAR file, but how to use it.